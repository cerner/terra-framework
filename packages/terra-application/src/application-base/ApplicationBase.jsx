import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';

import Base from 'terra-base';
import ThemeProvider from 'terra-theme-provider';
import { ActiveBreakpointProvider } from 'terra-breakpoints';

import styles from './ApplicationBase.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * The components to render within ApplicationBase.
   */
  children: PropTypes.node.isRequired,
  /**
   * The locale name to be used to load translated messages.
   */
  locale: PropTypes.string.isRequired,
  /**
   * Custom translations for the current locale.
   */
  customTranslatedMessages: (props, propName, componentName) => {
    if (!props[propName]) {
      return null;
    }

    if (Object.keys(props[propName]).length !== 0 && props.locale === undefined) {
      return new Error(`Missing locale prop for ${propName} in ${componentName} props`);
    }

    return null;
  },
  /**
   * The component to render while the translation files are being retrieved.
   * NOTE: Absolutely no locale-dependent logic should be
   * utilized in this placeholder.
   */
  translationsLoadingPlaceholder: PropTypes.node,
  /**
   * The name of the theme to apply to the application using terra-theme-provider.
   */
  themeName: PropTypes.string,
  /**
   * If provided, the theme styles are applied to the entire document.
   */
  themeIsGlobal: PropTypes.bool,
  /**
   * By default, the elements rendered by Application are fit to the Application's parent using 100% height.
   * If `fitToParentIsDisabled` is provided, the Application will render at its intrinsic content height and
   * overflow potentially overflow its parent.
   */
  fitToParentIsDisabled: PropTypes.bool,
};

const ApplicationBase = ({
  locale, customTranslatedMessages, translationsLoadingPlaceholder, themeName, themeIsGlobal, fitToParentIsDisabled, children,
}) => (
  <ThemeProvider
    className={cx('application-theme-provider', { fill: !fitToParentIsDisabled })}
    themeName={themeName}
    isGlobalTheme={themeIsGlobal}
  >
    <Base
      className={cx('application-base', { fill: !fitToParentIsDisabled })}
      customMessages={customTranslatedMessages}
      translationsLoadingPlaceholder={translationsLoadingPlaceholder}
      locale={locale}
    >
      <ActiveBreakpointProvider>
        {children}
      </ActiveBreakpointProvider>
    </Base>
  </ThemeProvider>
);

ApplicationBase.propTypes = propTypes;

export default ApplicationBase;
