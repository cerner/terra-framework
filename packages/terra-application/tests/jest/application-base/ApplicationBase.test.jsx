import React from 'react';

import ApplicationBase from '../../../src/application-base/ApplicationBase';

describe('Application', () => {
  it('should render with minimal props', () => {
    const wrapper = shallow((
      <ApplicationBase locale="en">
        <div>content</div>
      </ApplicationBase>
    ));
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with all props', () => {
    const wrapper = shallow((
      <ApplicationBase
        locale="en"
        customTranslatedMessages={{ custom: 'messages' }}
        translationsLoadingPlaceholder={<div>placeholder</div>}
        themeName="test-theme"
        themeIsGlobal
        fitToParentIsDisabled
      >
        <div>content</div>
      </ApplicationBase>
    ));
    expect(wrapper).toMatchSnapshot();
  });
});
