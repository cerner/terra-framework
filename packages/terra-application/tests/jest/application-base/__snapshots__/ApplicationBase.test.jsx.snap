// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application should render with all props 1`] = `
<ThemeProvider
  className="application-theme-provider"
  isGlobalTheme={true}
  themeName="test-theme"
>
  <Base
    className="application-base"
    customMessages={
      Object {
        "custom": "messages",
      }
    }
    locale="en"
    strictMode={false}
    translationsLoadingPlaceholder={
      <div>
        placeholder
      </div>
    }
  >
    <ActiveBreakpointProvider>
      <div>
        content
      </div>
    </ActiveBreakpointProvider>
  </Base>
</ThemeProvider>
`;

exports[`Application should render with minimal props 1`] = `
<ThemeProvider
  className="application-theme-provider fill"
  isGlobalTheme={false}
>
  <Base
    className="application-base fill"
    customMessages={Object {}}
    locale="en"
    strictMode={false}
  >
    <ActiveBreakpointProvider>
      <div>
        content
      </div>
    </ActiveBreakpointProvider>
  </Base>
</ThemeProvider>
`;
